experimental/e2e/storage/har.go

```diff
@@ -0,0 +1,260 @@
+			Title:           "Grafana E2E",
+			ID:              s.newUUID(),
+			PageTimings:     &har.PageTimings{},
+		}},
+	}
+}
+
+// Add converts the http.Request and http.Response to a har.Entry and adds it to the Fixture.
+func (s *HARStorage) Add(req *http.Request, res *http.Response) {
+	var (
+		err     error
+		reqBody []byte
+		resBody []byte
+	)
+
+	reqHeaders := make([]*har.NameValuePair, 0)
+	for name, value := range req.Header.Clone() {
+		reqHeaders = append(reqHeaders, &har.NameValuePair{Name: name, Value: value[0]})
+	}
+
+	resHeaders := make([]*har.NameValuePair, 0)
+	for name, value := range res.Header.Clone() {
+		resHeaders = append(resHeaders, &har.NameValuePair{Name: name, Value: value[0]})
+	}
+
+	queryString := make([]*har.NameValuePair, 0)
+	for name, value := range req.URL.Query() {
+		queryString = append(queryString, &har.NameValuePair{Name: name, Value: value[0]})
+	}
+
+	if req.Body != nil {
+		reqBody, err = utils.ReadRequestBody(req)
+		if err != nil {
+			fmt.Println("Failed to read request body", "err", err)
+		}
+	}
+
+	if res.Body != nil {
+		resBody, err = io.ReadAll(res.Body)
+		if err != nil {
+			fmt.Println("Failed to read response body", "err", err)
+		}
+		res.Body = ioutil.NopCloser(bytes.NewReader(resBody))
+	}
+
+	reqCookies := make([]*har.Cookie, 0)
+	for _, cookie := range req.Cookies() {
+		reqCookies = append(reqCookies, &har.Cookie{Name: cookie.Name, Value: cookie.Value})
+	}
+
+	resCookies := make([]*har.Cookie, 0)
+	for _, cookie := range res.Cookies() {
+		resCookies = append(resCookies, &har.Cookie{Name: cookie.Name, Value: cookie.Value})
+	}
+
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	s.har.Log.Entries = append(s.har.Log.Entries, &har.Entry{
+		StartedDateTime: s.currentTime().Format(time.RFC3339),
+		Time:            0.0,
+		Comment:         s.newUUID(),
+		Cache: &har.Cache{
+			Comment: "Not cached",
+		},
+		Timings: &har.Timings{
+			Send:    0.0,
+			Wait:    0.0,
+			Receive: 0.0,
+		},
+		Request: &har.Request{
+			Method:      req.Method,
+			HTTPVersion: req.Proto,
+			URL:         req.URL.String(),
+			Headers:     reqHeaders,
+			QueryString: queryString,
+			Cookies:     reqCookies,
+			BodySize:    int64(len(reqBody)),
+			PostData: &har.PostData{
+				MimeType: req.Header.Get("Content-Type"),
+				Text:     string(reqBody),
+			},
+		},
+		Response: &har.Response{
+			Status:      int64(res.StatusCode),
+			StatusText:  res.Status,
+			HTTPVersion: res.Proto,
+			Headers:     resHeaders,
+			HeadersSize: -1,
+			BodySize:    int64(len(resBody)),
+			Cookies:     resCookies,
+			RedirectURL: res.Header.Get("Location"),
+			Content: &har.Content{
+				Size:     int64(len(resBody)),
+				MimeType: res.Header.Get("Content-Type"),
+				Text:     string(resBody),
+			},
+		},
+	})
+}
+
+// Entries converts HAR entries to a slice of Entry (http.Request and http.Response pairs).
+func (s *HARStorage) Entries() []*Entry {
+	entries := make([]*Entry, len(s.har.Log.Entries))
+	s.lock.Lock()
```

Comment:
Just curious, the data is mostly written to which is why you opted to a Mutex (as opposed to a `RWMutex`?

Reply:
good call. switched to read lock here in https://github.com/grafana/grafana-plugin-sdk-go/pull/467